Resolve Smoke Test Issues

ronaldheft-usds opened the pull request:
**Why**

The smoke test is currently failing in Jenkins. These changes attempt to correct this issues with running the smoke test.

**What Changed**

* Removes failure time constraint from Taurus passfail module.
  * When tests were quickly failing, like is in the case of DEV due to BFD peering, the failure scenario was not being triggered.
* Adds a prompt when running local smoke tests to ensure authentication is enabled.
  * The API requires a token to derive the organizationID that is sent to the attribution server. If authentication is disabled, the API sends an empty organizationID to the attribution server and the tests fail.
* Links with https://github.com/CMSgov/dpc-ops/pull/69 to re-install virtualenv on Jenkins.

**Choices Made**

**Tickets closed**:

**Future Work**

**Checklist**

- [x] All tests are passing via `make ci-app` (app change) and `make ci-web` (website change)
- [x] Swagger documentation has been updated
- [x] FHIR documentation has been updated
- [x] Any required dpc-ops changes have a PR submitted and mentioned in this ticket
- [x] Any manual migration steps are documented, scripts written (where applicable), and tested
- [x] Before merging, any required dpc-ops changes have been approved and merged into master of the dpc-ops repo

ronaldheft-usds commented:
This is a work in progress, as submitting patients via the smoke test is still failing. Seeing the following exception from the smoke test:

```
2019-11-05 16:50:36,150 ERROR o.a.j.t.JMeterThread: Error while processing sampler: 'Java Request'.
java.lang.RuntimeException: Cannot submit patients.
        at gov.cms.dpc.api.client.ClientUtils.submitPatients(ClientUtils.java:285) ~[dpc-smoketest.jar:?]
        at gov.cms.dpc.testing.smoketests.SmokeTest.runTest(SmokeTest.java:103) ~[dpc-smoketest.jar:?]
        at org.apache.jmeter.protocol.java.sampler.JavaSampler.sample(JavaSampler.java:199) ~[ApacheJMeter_java.jar:5.1.1 r1855137]
        at org.apache.jmeter.threads.JMeterThread.doSampling(JMeterThread.java:622) ~[ApacheJMeter_core.jar:5.1.1 r1855137]
        at org.apache.jmeter.threads.JMeterThread.executeSamplePackage(JMeterThread.java:546) ~[ApacheJMeter_core.jar:5.1.1 r1855137]
        at org.apache.jmeter.threads.JMeterThread.processSampler(JMeterThread.java:486) [ApacheJMeter_core.jar:5.1.1 r1855137]
        at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:253) [ApacheJMeter_core.jar:5.1.1 r1855137]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: ca.uhn.fhir.rest.server.exceptions.InternalErrorException: HTTP 500 Internal Server Error
        at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
        at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
        at ca.uhn.fhir.rest.server.exceptions.BaseServerResponseException.newInstance(BaseServerResponseException.java:302) ~[dpc-smoketest.jar:?]
        at ca.uhn.fhir.rest.client.impl.BaseClient.invokeClient(BaseClient.java:351) ~[dpc-smoketest.jar:?]
        at ca.uhn.fhir.rest.client.impl.GenericClient$BaseClientExecutable.invoke(GenericClient.java:431) ~[dpc-smoketest.jar:?]
        at ca.uhn.fhir.rest.client.impl.GenericClient$OperationInternal.execute(GenericClient.java:1160) ~[dpc-smoketest.jar:?]
        at gov.cms.dpc.api.client.ClientUtils.bundleSubmitter(ClientUtils.java:274) ~[dpc-smoketest.jar:?]
        at gov.cms.dpc.api.client.ClientUtils.submitPatients(ClientUtils.java:283) ~[dpc-smoketest.jar:?]
        ... 7 more
```

On the attribution server side:

```
WARN  [2019-11-05 21:50:36,133 - dw-22 - POST /v1/Patient/$submit] org.hibernate.action.internal.UnresolvedEntityInsertActions: HHH000437: Attempting to save one or more entities that have a non-nullable association with an unsaved! org.hibernate.TransientPropertyValueException: Not-null property references a transient value - transient instance must be saved before current operation : gov.cms.dpc.common.entities.PatientEntity.organization -> gov.cms.dpc.comorg.hibernate.TransientPropertyValueException: Not-null property references a transient value - transient instance must be saved before current operation : gov.cms.dpc.common.entities.PatientEntity.organization -> gov.cms.dpc.common.entities.OrganizationEntity
        at org.hibernate.action.internal.UnresolvedEntityInsertActions.checkNoUnresolvedActionsAfterOperation(UnresolvedEntityInsertActions.java:122)
        at org.hibernate.engine.spi.ActionQueue.checkNoUnresolvedActionsAfterOperation(ActionQueue.java:436)
        at org.hibernate.internal.SessionImpl.checkNoUnresolvedActionsAfterOperation(SessionImpl.java:569)
        at org.hibernate.internal.SessionImpl.fireSaveOrUpdate(SessionImpl.java:599)
        at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:591)
        at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:586)
        at io.dropwizard.hibernate.AbstractDAO.persist(AbstractDAO.java:202)
        at gov.cms.dpc.attribution.jdbi.PatientDAO.persistPatient(PatientDAO.java:22)
        at gov.cms.dpc.attribution.resources.v1.PatientResource.createPatient(PatientResource.java:121)
        at gov.cms.dpc.attribution.utils.RESTUtils.lambda$bulkResourceHandler$1(RESTUtils.java:43)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at gov.cms.dpc.attribution.utils.RESTUtils.bulkResourceHandler(RESTUtils.java:51)
        at gov.cms.dpc.attribution.resources.v1.PatientResource.bulkSubmitPatients(PatientResource.java:136)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
        at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
        at io.dropwizard.jetty.NonblockingServletHolder.handle(NonblockingServletHolder.java:49)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
        at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:45)
        at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:39)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:89)
        at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120)
        at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:135)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:239)
        at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:703)
        at io.dropwizard.jetty.BiDiGzipHandler.handle(BiDiGzipHandler.java:67)
        at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:56)
        at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:174)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:505)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

This appears to me like the issue is the smoke test is missing the `ManagingOrganization`, since the Postman end to end tests are able to run successfully against this Attribution endpoint through the API. I've attempted to add the `ManagingOrganization` in this PR (the same way the API does), but I need to debug more to verify that's actually being added.

A similar issue was occurring on the `Practitioner` resource, which was resolved by adding the `organizationID` to bundle.

nickrobison-usds commented:
Is this error from Dev? Tests are failing in dev because the certificate doesn’t match and HAPI throws an exception.

ronaldheft-usds commented:
@nickrobison-usds Nope, this is locally. Are the latest smoke tests in master?

nickrobison-usds commented:
They are, and they were passing on my machine before I merged them in.

ronaldheft-usds commented:
@nickrobison-usds Yeah, the tests are passing in master, but Taurus is reading 0% success as a pass through virtualenv. When I looked into the tests more closely, I saw they were erroring and being swallowed in the Python virtualenv. When I ran them directly, I saw the attribution server was throwing 500 exceptions, because the smoke tests were calling the endpoints without the `organizationID`.

The smoke tests are calling the attribution server directly, correct? And they're not currently doing anything to add the organizationID, correct, or am I missing that somewhere?

ronaldheft-usds commented:
@nickrobison-usds Got this squared away. The issue with running the tests locally turned out to be pretty simple. See this in the changelog:

* Adds a prompt when running local smoke tests to ensure authentication is enabled.
  * The API requires a token to derive the organizationID that is sent to the attribution server. If authentication is disabled, the API sends an empty organizationID to the attribution server and the tests fail.

Rebased and removed all the junk I had added trying to figure out the source of this issue.

ronaldheft-usds commented:
@nickrobison-usds Also, the reason the tests were passing in DEV, was not due to virtualenv, but the time constraint put on a failures. It was confusing, because the output was showing FAILED, but the exit code was 0. Virtualenv does pass through the exit code when the correct exit code is returned.
